cmd/server/main.go

```diff
@@ -155,3 +161,107 @@ func setupLogger() *zap.SugaredLogger {
 	}
 	return zapLogger.Sugar()
 }
+
+func leaderElection() {
+	log.Println("Trying leader election")
+
+	var leaseLockName = "testlease"
+	var leaseLockNamespace = "newrelic"
+	var id = os.Getenv("HOSTNAME")
+
+	// leader election uses the Kubernetes API by writing to a
+	// lock object, which can be a LeaseLock object (preferred),
+	// a ConfigMap, or an Endpoints (deprecated) object.
+	// Conflicting writes are detected and each client handles those actions
+	// independently.
+	client := NewKubernetes()
+	log.Println("We hope we created the client")
+
+	run := func(ctx context.Context) {
+		// complete your controller loop here
+
+		log.Println("Controller loop...")
+
+		tick := time.Tick(time.Second)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-tick:
+				fmt.Println("I am the leader apparently")
+			}
+		}
+	}
+
+	// use a Go context so we can tell the leaderelection code when we
+	// want to step down
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// listen for interrupts or the Linux SIGTERM signal and cancel
+	// our context, which the leader election code will observe and
+	// step down
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-ch
+		log.Println("Received termination, signaling shutdown")
+		cancel()
+	}()
+
+	// we use the Lease lock type since edits to Leases are less common
+	// and fewer objects in the cluster watch "all Leases".
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      leaseLockName,
+			Namespace: leaseLockNamespace,
+		},
+		Client: client.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+	}
+
+	// start the leader election code loop
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock: lock,
+		// IMPORTANT: you MUST ensure that any code you have that
+		// is protected by the lease must terminate **before**
+		// you call cancel. Otherwise, you could have a background
+		// loop still running and another process could
+		// get elected before your background loop finished, violating
+		// the stated goal of the lease.
+		ReleaseOnCancel: true,
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   15 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				// we're notified when we start - this is where you would
+				// usually put your code
+				run(ctx)
+			},
+			OnStoppedLeading: func() {
+				// we can do cleanup here
+				log.Printf("leader lost: %s", id)
+				os.Exit(0)
```

Comment:
I think we could just stop the controller loop (which we do not have right now) instead of exiting process, so we can re-start it once we get leadership again.